pulsar-broker/src/test/java/org/apache/pulsar/broker/service/TransactionMetadataStoreServiceTest.java

```diff
@@ -222,14 +233,58 @@ public void testTimeoutTrackerMultiThreading() throws Exception {
             int i = -1;
             while (++i < 100) {
                 try {
-                    transactionMetadataStore.newTransaction(10000);
+                    transactionMetadataStore.newTransaction(4000);
                 } catch (Exception e) {
                     //no operation
                 }
             }
         }).start();
-        Awaitility.await().atLeast(3000, TimeUnit.MICROSECONDS).atMost(10000, TimeUnit.MILLISECONDS)
-                .until(() -> txnMap.size() == 100);
+
+        checkoutTimeout(txnMap, 300);
+        checkoutTimeout(txnMap, 200);
+        checkoutTimeout(txnMap, 100);
+        checkoutTimeout(txnMap, 0);
+    }
+
+    private void checkoutTimeout(ConcurrentMap<TxnID, Pair<TxnMeta, List<Position>>> txnMap, int time) {
+        Awaitility.await().atLeast(1000, TimeUnit.MICROSECONDS).atMost(2000, TimeUnit.MILLISECONDS)
+                .until(() -> txnMap.size() == time);
+    }
+
+    @Test
+    public void transactionTimeoutRecoverTest() throws Exception {
+        int timeout = 2000;
+        pulsar.getTransactionMetadataStoreService().addTransactionMetadataStore(TransactionCoordinatorID.get(0));
+        Awaitility.await().atMost(2000, TimeUnit.MILLISECONDS)
+                .until(() -> pulsar.getTransactionMetadataStoreService()
+                        .getStores().get(TransactionCoordinatorID.get(0)) != null);
+        MLTransactionMetadataStore transactionMetadataStore =
+                (MLTransactionMetadataStore) pulsar.getTransactionMetadataStoreService()
+                        .getStores().get(TransactionCoordinatorID.get(0));
+        checkTransactionMetadataStoreReady(transactionMetadataStore);
+
+        transactionMetadataStore.newTransaction(timeout);
+
+        pulsar.getTransactionMetadataStoreService()
+                .removeTransactionMetadataStore(TransactionCoordinatorID.get(0));
+
+        pulsar.getTransactionMetadataStoreService().addTransactionMetadataStore(TransactionCoordinatorID.get(0));
+        Awaitility.await().atMost(2000, TimeUnit.MILLISECONDS)
+                .until(() -> pulsar.getTransactionMetadataStoreService()
+                        .getStores().get(TransactionCoordinatorID.get(0)) != null);
+        transactionMetadataStore =
+                (MLTransactionMetadataStore) pulsar.getTransactionMetadataStoreService()
+                        .getStores().get(TransactionCoordinatorID.get(0));
+
+        checkTransactionMetadataStoreReady(transactionMetadataStore);
+
+        Field field = MLTransactionMetadataStore.class.getDeclaredField("txnMetaMap");
+        field.setAccessible(true);
+        ConcurrentMap<TxnID, Pair<TxnMeta, List<Position>>> txnMap =
+                (ConcurrentMap<TxnID, Pair<TxnMeta, List<Position>>>) field.get(transactionMetadataStore);
+        assertEquals(txnMap.size(), 1);
```

Comment:
@congbobo184  This is flaky, could it be removed?

Reply:
+1 for removing it.

@lhotari  can you please send a PR ?

Reply:
@eolivelli Done, #10313